### PR TITLE
Add MPI support

### PR DESCRIPTION
**Pull request by shadowwalkersb**

### CODE REVIEW

**Comment by shadowwalkersb** [comment]
Will be adding tests after #24 is merged. Otherwise, this should be ready for review.

**Comment by jakirkham** [comment]
So there has been some discussion about how to handle `mpi`. I think it will be feature based and will use some metapackage. This will also probably need to move to a matrix item or another branch if it won't fit in CI limits reasonably.

cc @minrk @dalcinl

**Comment by shadowwalkersb** [comment]
Can we merge this PR, so we have `fftw-mpi`? We can incorporate the `mpi` metapackage once it is implemented. The builds succeeded, no time out. Of course, let's merge #24 first.

**Comment by jakirkham** [comment]
TL;DR While I'm generally ok with adding optional MPI support here, I'm not willing to merge this as is.

First there are people that have been working on this problem for a while. I'd like to hear what they have to say and incorporate their feedback. This PR hasn't been open that long and this is a busy place with people doing other things outside of here. Waiting for their feedback is not unreasonable. As they have been gaining expertise in this area and are interested in healthy MPI ecosystem at conda-forge, I think there is a lot to be gained by hearing from them.

Second there are several choices for MPI. People should have freedom to chose the MPI implementation if they have a preference. Also they should have freedom to choose none at all if they don't want it. This PR does not provide these options in its current form. So I'd like to see it revised to handle these options somehow. This is where the discussion and feedback from others will be crucial.

So let's wait and see what they have to say. Then we can determine the best path forward.

**Comment by shadowwalkersb** [comment]
All sounds good.

Is the new `variants` concept helpful here?

**Comment by minrk** [comment]
@shadowwalkersb I don't think variants are available yet in the version of conda-build used on conda-forge. Even so, I'm not sure they solve this particular case. I'm looking into it, though.

@jakirkham I have an implementation I'd like to try for MPI variants, but haven't gotten an OK from conda-forge folks in https://github.com/conda-forge/staged-recipes/pull/1501. If you or someone sign off, I can get started rolling out MPI variants and see how it goes. The one that seems to work best does not involve a metapackage.

One case that isn't handled in the mpi variant proposal is for tools like hdf5 or fftw that want "mpi or not", as opposed to "mpich or openmpi". I think it would not serve users well to follow the blas example and have a 'nompi' variant of an mpi metapackage, because it seems like installing the no-mpi version of fftw shouldn't prevent the installation of other mpi-using libraries. But maybe that's not a big deal, given the use of envs.

Similarly, there does not appear to be any mechanism for packages to express a variant preference, e.g. `fftw` being installed without MPI by default.

**Comment by smithsp** [comment]
Thanks for this pull request, as it helped me move forward with my own MPI build.

**Comment by shadowwalkersb** [comment]
Closing as outdated.
